sync/sync.go

```diff
@@ -1,81 +1,103 @@
 package sync
 
 import (
-	"errors"
+	"context"
 	"log"
-	"sync/atomic"
+	"runtime"
 
 	"github.com/NethermindEth/juno/blockchain"
+	"github.com/NethermindEth/juno/core"
 	"github.com/NethermindEth/juno/starknetdata"
+	"github.com/sourcegraph/conc/stream"
 )
 
 // Synchronizer manages a list of StarknetData to fetch the latest blockchain updates
 type Synchronizer struct {
-	running uint64
-
 	Blockchain   *blockchain.Blockchain
 	StarknetData starknetdata.StarknetData
-
-	quit chan struct{}
 }
 
 func NewSynchronizer(bc *blockchain.Blockchain, starkNetData starknetdata.StarknetData) *Synchronizer {
 	return &Synchronizer{
-		running:      0,
 		Blockchain:   bc,
 		StarknetData: starkNetData,
-		quit:         make(chan struct{}),
 	}
 }
 
 // Run starts the Synchronizer, returns an error if the loop is already running
-func (s *Synchronizer) Run() error {
-	if running := atomic.CompareAndSwapUint64(&s.running, 0, 1); !running {
-		return errors.New("synchronizer is already running")
-	}
-	return s.SyncBlocks()
+func (s *Synchronizer) Run(ctx context.Context) error {
+	return s.SyncBlocks(ctx)
 }
 
-// Shutdown attempts to stop the Synchronizer, should block until loop acknowledges the request
-func (s *Synchronizer) Shutdown() error {
-	if stopped := atomic.CompareAndSwapUint64(&s.running, 1, 0); !stopped {
-		return errors.New("synchronizer is stopped")
+func (s *Synchronizer) SyncBlocks(ctx context.Context) error {
+	fetchers := stream.New().WithMaxGoroutines(runtime.NumCPU())
+	verifiers := stream.New().WithMaxGoroutines(runtime.NumCPU())
```

Comment:
What will happen if we were to submit max goroutines higher than the number of CPUs?

Reply:
Not much, we would overwhelm our CPU and some of the goroutines would have to wait to be scheduled on a thread. So sweet spot is having exactly NumCPU number of computation heavy goroutines.

Reply:
Not much, we would overwhelm our CPU and some of the goroutines would have to wait to be scheduled on a thread. So sweet spot is having exactly NumCPU number of computation heavy goroutines.

Reply:
Shouldn't be `runtime.NumCPU()/2`? Given we have 2 streams?

Reply:
fetchers are not computation heavy routines. they are mostly blocked, waiting for network.